### PR TITLE
HTTP1: Add DumpState support for HTTP1.

### PR DESCRIPTION
**Pull request by KBaichoo**
Signed-off-by: Kevin Baichoo <kbaichoo@google.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)

Commit Message: HTTP1: Add DumpState support for HTTP1.
Additional Description: While dispatching, the HTTP1 Connection object registers itself as a `scopetrackedobject`. If we crash during parsing pipeline, we'll dump information such as the partial headers parsed, current buffer, headermap, etc.
Risk Level: medium
Testing: unit tests
Docs Changes: N/A
Release Notes: Included
Platform Specific Features: N/A
Related Issue #14249

/assign @antoniovicente

### CODE REVIEW

File: source/common/http/http1/codec_impl.cc [hunk 1]
```diff
@@ -861,6 +864,58 @@ void ConnectionImpl::onResetStreamBase(StreamResetReason reason) {
...
+
+  // Dump the first slice of the dispatching buffer and buffered_body if
+  // applicable.
+  auto dumpBuffer = [](auto* instance) {
+    auto slice = instance->frontSlice();
+    return absl::string_view(static_cast<const char*>(slice.mem_), slice.len_);
+  };
+
+  os << DUMP_NULLABLE_MEMBER(current_dispatching_buffer_, dumpBuffer(current_dispatching_buffer_));
+  os << DUMP_MEMBER(buffered_body_, dumpBuffer(&buffered_body_));
```
Comment (antoniovicente) [inline source/common/http/http1/codec_impl.cc:883]: I think we shoud consider dumping just buffered_body_.length() or omit this field, the body contents is not that relevant. The current_dispatching_buffer_ seems more relevant.

File: source/common/http/http1/codec_impl.cc [hunk 2]
```diff
@@ -861,6 +864,58 @@ void ConnectionImpl::onResetStreamBase(StreamResetReason reason) {
...
+  const char* spaces = spacesForLevel(indent_level);
+  // Dump all bool to provide context
+  os << spaces << "Http1::ConnectionImpl " << this << DUMP_MEMBER(dispatching_)
+     << DUMP_MEMBER(dispatching_slice_already_drained_) << DUMP_MEMBER(reset_stream_called_)
+     << DUMP_MEMBER(handling_upgrade_) << DUMP_MEMBER(deferred_end_stream_headers_)
+     << DUMP_MEMBER(strict_1xx_and_204_headers_) << DUMP_MEMBER(processing_trailers_);
+
+  // Dump the first slice of the dispatching buffer and buffered_body if
+  // applicable.
+  auto dumpBuffer = [](auto* instance) {
```
Comment (antoniovicente) [inline source/common/http/http1/codec_impl.cc:877]: dumpFrontSlice

File: source/common/http/http1/codec_impl.cc [hunk 3]
```diff
@@ -861,6 +864,58 @@ void ConnectionImpl::onResetStreamBase(StreamResetReason reason) {
...
+    auto slice = instance->frontSlice();
```
Comment (antoniovicente) [inline source/common/http/http1/codec_impl.cc:878]: Should you quote and escape the buffer contents?  Should you explicitly dump the size of the slice?

File: source/common/http/http1/codec_impl.cc [hunk 4]
```diff
@@ -861,6 +864,58 @@ void ConnectionImpl::onResetStreamBase(StreamResetReason reason) {
...
+    return absl::string_view(static_cast<const char*>(slice.mem_), slice.len_);
+  };
+
+  os << DUMP_NULLABLE_MEMBER(current_dispatching_buffer_, dumpBuffer(current_dispatching_buffer_));
```
Comment (antoniovicente) [inline source/common/http/http1/codec_impl.cc:882]: Parsing is a destructive operation on current_dispatching_buffer_ in some cases involving processing bodies.  I guess this may dump the wrong slice in those cases.  That said, crashing while processing body is unlikely; crashes during header parse should produce the right output.

File: source/common/http/http1/codec_impl.cc [hunk 5]
```diff
@@ -861,6 +864,58 @@ void ConnectionImpl::onResetStreamBase(StreamResetReason reason) {
...
+}
+
+void ServerConnectionImpl::dumpAdditionalState(std::ostream& os, int indent_level) const {
+  const char* spaces = spacesForLevel(indent_level);
+  // Dump header map, it may be null if it was moved to the request.
+  if (absl::holds_alternative<RequestHeaderMapPtr>(headers_or_trailers_)) {
+    DUMP_DETAILS(absl::get<RequestHeaderMapPtr>(headers_or_trailers_));
+  } else {
+    DUMP_DETAILS(absl::get<RequestTrailerMapPtr>(headers_or_trailers_));
+  }
```
Comment (antoniovicente) [inline source/common/http/http1/codec_impl.cc:906]: Please consider dumping ActiveRequest::request_url_;

File: source/common/http/http1/codec_impl.cc [hunk 6]
```diff
@@ -861,6 +864,58 @@ void ConnectionImpl::onResetStreamBase(StreamResetReason reason) {
...
+}
+
+void ClientConnectionImpl::dumpAdditionalState(std::ostream& os, int indent_level) const {
```
Comment (antoniovicente) [inline source/common/http/http1/codec_impl.cc:909]: Should this be an override of dumpState instead?

File: source/common/http/http1/codec_impl.cc [hunk 7]
```diff
@@ -861,6 +864,58 @@ void ConnectionImpl::onResetStreamBase(StreamResetReason reason) {
...
+  os << spaces << "Http1::ConnectionImpl " << this << DUMP_MEMBER(dispatching_)
+     << DUMP_MEMBER(dispatching_slice_already_drained_) << DUMP_MEMBER(reset_stream_called_)
+     << DUMP_MEMBER(handling_upgrade_) << DUMP_MEMBER(deferred_end_stream_headers_)
+     << DUMP_MEMBER(strict_1xx_and_204_headers_) << DUMP_MEMBER(processing_trailers_);
+
+  // Dump the first slice of the dispatching buffer and buffered_body if
+  // applicable.
+  auto dumpBuffer = [](auto* instance) {
+    auto slice = instance->frontSlice();
+    return absl::string_view(static_cast<const char*>(slice.mem_), slice.len_);
```
Comment (antoniovicente) [inline source/common/http/http1/codec_impl.cc:879]: Excellent, I see tests that cover dumping of current_dispatching_buffer_ in 
Http1ClientConnectionImplTest.ShouldDumpBuffersWithoutAllocatingMemory.

File: source/common/http/http1/codec_impl.cc [hunk 8]
```diff
@@ -861,6 +864,58 @@ void ConnectionImpl::onResetStreamBase(StreamResetReason reason) {
...
+  // Dump header parsing state, and any progress on other headers.
+  os << DUMP_MEMBER(header_parsing_state_);
+  if (header_parsing_state_ == HeaderParsingState::Field) {
+    os << DUMP_MEMBER(current_header_field_, current_header_field_.getStringView());
+  } else if (header_parsing_state_ == HeaderParsingState::Value) {
+    os << DUMP_MEMBER(current_header_field_, current_header_field_.getStringView());
+    os << DUMP_MEMBER(current_header_value_, current_header_value_.getStringView());
+  }
+
+  // Dump Child
```
Comment (antoniovicente) [inline source/common/http/http1/codec_impl.cc:894]: I think that the dump for current_dispatching_buffer_ should be done via DUMP_DETAILS.  Consider putting it after the dump of header maps in subclasses. Reasoning: buffer contents can be long and may contain newlines

**Review by antoniovicente — COMMENTED** [review]
Thanks for further improving debug dumps when crashing in H1 parsing.

File: source/common/http/http1/codec_impl.cc (hunk 4 above)
Comment (KBaichoo) [reply source/common/http/http1/codec_impl.cc]: Will not dump the first slice then if the dispatching buffer is already drained.

**Review by KBaichoo — COMMENTED** [review]
Thanks for the review Antonio.

File: source/common/http/http1/codec_impl.cc (hunk 3 above)
Comment (KBaichoo) [reply source/common/http/http1/codec_impl.cc]: The buffer contents are escaped in `\t, \r, \n, ', " and \`

File: source/common/http/http1/codec_impl.cc [hunk 9]
```diff
@@ -861,6 +864,79 @@ void ConnectionImpl::onResetStreamBase(StreamResetReason reason) {
...
+    // Use the macro for consistent formatting.
+    os << DUMP_NULLABLE_MEMBER(current_dispatching_buffer_, "null");
+    return;
+  } else {
+    auto front_slice = [](Buffer::Instance* instance) {
+      auto slice = instance->frontSlice();
+      return absl::string_view(static_cast<const char*>(slice.mem_), slice.len_);
+    }(current_dispatching_buffer_);
+
+    os << spaces << DUMP_MEMBER(front_slice.length(), front_slice.length()) << ", front_slice: \n";
```
Comment (antoniovicente) [inline source/common/http/http1/codec_impl.cc:901]: I would consider doing the following here:

```
os << spaces << "current_dispatching_buffer_ front_slice length: " << front_slice.length()
   << " contents: \"" <<  StringUtil::escapeToOstream(os, front_slice) << "\"\n";
```

File: test/common/http/http1/codec_impl_test.cc [hunk 10]
```diff
@@ -1913,6 +1917,76 @@ TEST_F(Http1ServerConnectionImplTest, TestSmugglingAllowChunkedContentLength100)
...
+      .WillOnce(Invoke([&](ResponseEncoder&, bool) -> RequestDecoder& { return decoder; }));
+
+  std::array<char, 1024> buffer;
+  OutputBufferStream ostream{buffer.data(), buffer.size()};
+
+  // Dump the body
+  Buffer::OwnedImpl request("POST / HTTP/1.1\r\n"
+                            "Content-Length: 11\r\n"
+                            "\r\n"
+                            "Hello Envoy");
```
Comment (antoniovicente) [inline test/common/http/http1/codec_impl_test.cc:1969]: Would it be possible to split the request over 2 slices and verify that only the first slice is logged?

Something like:
```
Buffer::OwnedImpl request;
request.appendSliceForTest("POST / HTTP/1.1\r\n"
                            "Content-Length: 11\r\n"
                            "\r\n"
                            "Hello");
request.appendSliceForTest(" Envoy");

  EXPECT_THAT(ostream.contents(),
￼             HasSubstr("front_slice.length(): 44 ...
```

File: source/common/http/http1/codec_impl.cc [hunk 11]
```diff
@@ -861,6 +864,79 @@ void ConnectionImpl::onResetStreamBase(StreamResetReason reason) {
...
 }
 
+void ConnectionImpl::dumpState(std::ostream& os, int indent_level) const {
+  const char* spaces = spacesForLevel(indent_level);
+  // Dump all bool to provide context
+  os << spaces << "Http1::ConnectionImpl " << this << DUMP_MEMBER(dispatching_)
+     << DUMP_MEMBER(dispatching_slice_already_drained_) << DUMP_MEMBER(reset_stream_called_)
+     << DUMP_MEMBER(handling_upgrade_) << DUMP_MEMBER(deferred_end_stream_headers_)
+     << DUMP_MEMBER(strict_1xx_and_204_headers_) << DUMP_MEMBER(processing_trailers_)
+     << DUMP_MEMBER(buffered_body_.length(), buffered_body_.length());
```
Comment (antoniovicente) [inline source/common/http/http1/codec_impl.cc:874]: Doesn't this work?

<< DUMP_MEMBER(buffered_body_.length())

**Review by antoniovicente — COMMENTED** [review]
Just a few nits, thanks again for improving crash dump coverage.

File: test/common/http/http1/codec_impl_test.cc (hunk 10 above)
Comment (KBaichoo) [reply test/common/http/http1/codec_impl_test.cc]: Did something similar to this; We were using onMessageCompleteBase's callback from the parser to dump the data of the buffer. 

I added extraneous bytes in the other slice to  check that only the front slice is dumped. The content length is used to tell the parser the message should only have the front slice.

**Review by KBaichoo — COMMENTED** [review]
Thanks for the review Antonio.

**Comment by antoniovicente** [comment]
grpc errors during build?

/retest

**Comment by KBaichoo** [comment]
Not sure if that restarted it; did run clang-tidy in docker locally no issue...

/retest

**Comment by KBaichoo** [comment]
grpc issue seems to be due to some loadshedding that some service is performing -- maybe too much going on for CI.

Will re-kick this in a bit, perhaps we should get some other reviewing eyes on it in the mean time?

**Comment by KBaichoo** [comment]
PTAL @envoyproxy/senior-maintainers; seems like it was LGTM from @antoniovicente, but he's OOO.

**Comment by jmarantz** [comment]
needs main merge.

File: source/common/http/http1/codec_impl.cc [hunk 12]
```diff
@@ -863,6 +866,76 @@ void ConnectionImpl::onResetStreamBase(StreamResetReason reason) {
...
+  if (current_dispatching_buffer_ == nullptr || dispatching_slice_already_drained_) {
+    // Buffer is either null or already drained (in the body).
+    // Use the macro for consistent formatting.
+    os << DUMP_NULLABLE_MEMBER(current_dispatching_buffer_, "drained");
+    return;
+  } else {
+    auto front_slice = [](Buffer::Instance* instance) {
+      auto slice = instance->frontSlice();
+      return absl::string_view(static_cast<const char*>(slice.mem_), slice.len_);
+    }(current_dispatching_buffer_);
```
Comment (antoniovicente) [inline source/common/http/http1/codec_impl.cc:901]: The previous 4 lines are hard to follow due to the inline lambda use and excessive use of 'auto'.

Consider instead:
```
absl::string_view front_slice = [](RawSlice slice) {
  return absl::string_view(static_cast<const char*>(slice.mem_), slice.len_);
}(current_dispatching_buffer_->frontSlice());
```

or

```
auto to_str_view = [](RawSlice slice) {
  return absl::string_view(static_cast<const char*>(slice.mem_), slice.len_);
};
absl::string_view front_slice = to_str_view(current_dispatching_buffer_->frontSlice());
```

File: source/common/common/dump_state_utils.h [hunk 13]
```diff
@@ -9,13 +9,25 @@ namespace Envoy {
...
+#define _DUMP_MEMBER_SELECTOR(_1, _2, DUMP_MACRO, ...) DUMP_MACRO
+
+// This is a workaround for fact that MSVC expands __VA_ARGS__ after passing them into a macro,
+// rather than before passing them into a macro. Without this,
+// _DUMP_MEMBER_SELECTOR does not work correctly when compiled with MSVC.
+#define EXPAND(X) X
+
+// If DUMP_MEMBER is called with one argument, then _DUMP_MEMBER is called.
+// If DUMP_MEMBER is called with two arguments, then _DUMP_MEMBER_VIA_VALUE is called.
+#define DUMP_MEMBER(...)                                                                           \
```
Comment (antoniovicente) [inline source/common/common/dump_state_utils.h:23]: Code would be easier to follow if you use a different macro name in cases we want to override how the member is printed.  Something like:
```
#define DUMP_MEMBER_VIA_VALUE(member, value) ", " #member ": " << (value)
```

Other possible macro names include DUMP_MEMBER_VIA and DUMP_MEMBER_AS

File: test/common/http/http1/codec_impl_test.cc [hunk 14]
```diff
@@ -2635,7 +2713,8 @@ TEST_F(Http1ClientConnectionImplTest, LowWatermarkDuringClose) {
 
   EXPECT_CALL(response_decoder, decodeHeaders_(_, true))
       .WillOnce(Invoke([&](ResponseHeaderMapPtr&, bool) {
-        // Fake a call for going below the low watermark. Make sure no stream callbacks get called.
+        // Fake a call for going below the low watermark. Make sure no stream callbacks get
+        // called.
```
Comment (antoniovicente) [inline test/common/http/http1/codec_impl_test.cc:2717]: Unclear why the previous 3 whitespace changes were done.  Was it the fix format script or a consequence of some indent change that was later backed out?

**Review by antoniovicente — COMMENTED** [review]
Some style nits, looks good otherwise.

File: source/common/http/http1/codec_impl.cc [hunk 15]
```diff
@@ -863,6 +866,76 @@ void ConnectionImpl::onResetStreamBase(StreamResetReason reason) {
...
+  } else if (header_parsing_state_ == HeaderParsingState::Value) {
+    os << DUMP_MEMBER(current_header_field_, current_header_field_.getStringView());
+    os << DUMP_MEMBER(current_header_value_, current_header_value_.getStringView());
+  }
+
+  // Dump Child
+  os << '\n';
+  dumpAdditionalState(os, indent_level);
+
+  // Dump the first slice of the dispatching buffer if not drained.
```
[the remark this answers is not in the document]
Comment (KBaichoo) [reply source/common/http/http1/codec_impl.cc:891]: Done.

File: source/common/http/http1/codec_impl.cc (hunk 12 above)
Comment (KBaichoo) [reply source/common/http/http1/codec_impl.cc:901]: went with the former.

File: source/common/common/dump_state_utils.h (hunk 13 above)
Comment (KBaichoo) [reply source/common/common/dump_state_utils.h:23]: sg. will go with `DUMP_MEMBER_AS`.

File: source/common/common/utility.cc [hunk 16]
```diff
@@ -457,6 +457,34 @@ std::string StringUtil::escape(const std::string& source) {
   return ret;
 }
 
+void StringUtil::escapeToOstream(std::ostream& os, absl::string_view view) {
```
[the remark this answers is not in the document]
Comment (KBaichoo) [reply source/common/common/utility.cc:460]: Will do in a followup if needed. Added a todo here.

File: source/common/http/http1/codec_impl.cc [hunk 17]
```diff
@@ -863,6 +866,76 @@ void ConnectionImpl::onResetStreamBase(StreamResetReason reason) {
...
+     << DUMP_MEMBER(buffered_body_.length());
+
+  // Dump header parsing state, and any progress on other headers.
+  os << DUMP_MEMBER(header_parsing_state_);
+  if (header_parsing_state_ == HeaderParsingState::Field) {
+    os << DUMP_MEMBER(current_header_field_, current_header_field_.getStringView());
+  } else if (header_parsing_state_ == HeaderParsingState::Value) {
+    os << DUMP_MEMBER(current_header_field_, current_header_field_.getStringView());
+    os << DUMP_MEMBER(current_header_value_, current_header_value_.getStringView());
+  }
```
[the remark this answers is not in the document]
Comment (KBaichoo) [reply source/common/http/http1/codec_impl.cc:885]: Done.

File: test/common/http/http1/codec_impl_test.cc (hunk 14 above)
Comment (KBaichoo) [reply test/common/http/http1/codec_impl_test.cc:2717]: The latter, I've reverted it.

**Review by KBaichoo — COMMENTED** [review]
Thanks

File: source/common/common/utility.cc [hunk 18]
```diff
@@ -457,6 +457,7 @@ std::string StringUtil::escape(const std::string& source) {
   return ret;
 }
 
+// TODO(kbaichoo): If needed, add support for escaping < and chars >= 127.
```
Comment (antoniovicente) [inline source/common/common/utility.cc:460]: escaping chars < 32 and >= 127.

File: source/common/http/http1/codec_impl.cc [hunk 19]
```diff
@@ -912,11 +907,11 @@ void ConnectionImpl::dumpState(std::ostream& os, int indent_level) const {
...
-  os << DUMP_MEMBER(active_request_.request_url_,
-                    active_request_.has_value() &&
-                            !active_request_.value().request_url_.getStringView().empty()
-                        ? active_request_.value().request_url_.getStringView()
-                        : "null\n");
+  os << DUMP_MEMBER_AS(active_request_.request_url_,
+                       active_request_.has_value() &&
+                               !active_request_.value().request_url_.getStringView().empty()
+                           ? active_request_.value().request_url_.getStringView()
+                           : "null\n");
```
Comment (antoniovicente) [inline source/common/http/http1/codec_impl.cc:914]: there's a trailing newline after "null", but no similar newline is added when the URL is not empty.

**Review by antoniovicente — APPROVED** [review]

File: source/common/http/http1/codec_impl.cc (hunk 19 above)
Comment (KBaichoo) [reply source/common/http/http1/codec_impl.cc]: Done, they both have the newline now.

**Review by antoniovicente — APPROVED** [review]

**Comment by KBaichoo** [comment]
PTAL @ggreenway since you're the maintainer on-call this week. Thanks!

**Comment by KBaichoo** [comment]
/retest 

Flaky integration test.

File: source/common/common/utility.cc (hunk 16 above)
[the remark this answers is not in the document]
Comment (ggreenway) [reply source/common/common/utility.cc:461]: There's also escaping code in `source/common/common/logger.cc` in `DelegatingLogSink::escapeLogLine()` which calls `absl::CEscape`. Can you use that instead?

[the remark this answers is not in the document]
Comment (KBaichoo) [reply source/common/common/utility.cc:461]: Nope, `absl::CEscape` allocs memory via the string it'll create and return. This avoids memory allocs.

**Review by snowp — APPROVED** [review]
Awesome, this looks super useful!
